doc/architecture-decisions/0018-route-crd-and-contour-controller.md

```diff
@@ -0,0 +1,48 @@
+# 18. Add Contour as an Alternative Ingress Solution Provider
+
+Date: 2020-10-09
+
+## Status
+
+Proposal
+
+## Context
+In our efforts to allow more optionality in ingress solutions for CF-for-K8s, we
+want to allow the use of Contour as a potential alternative to Istio.
+
+### Proposed Design
+
+We propose:
+1. Adding a configuration option `ingress_solution_provider` to CF-for-K8s, with
+   potential values `istio` or `contour`.
+2. Creating another kubebuilder controller, similar to routecontroller, to
+   handle when `ingress_solution_provider: contour` is configured.
+3. Renaming `Route Controller` to `Istio Controller`
+
+![Proposed ContourControllerDesign](../assets/contour-controller.png)
+
+### Open Questions
+1. What do we do about the config? Does contour config live in cf-k8s-networking
+   and eventually it moves to cf-for-k8s?
+2. What happens if an operator wants to change their ingress solution provider,
+   will they have to redeploy CF-for-K8s? Is that a big deal?
+3. Looks like Contour deploys Envoys as a daemonset. Is that gonna be a problem?
+
+## Decision
+Waiting on Review
+
+## Consequences
+* We will have two separate controllers. When an operator makes a decision by
+  configuring `ingress_solution_provider`, only one should be
+  deployed/active at a time.
+* Because only one controller is active at a time, the available networking
```

Comment:
Isn't one of the biggest consequences for cf-for-k8s as a whole that we loose automatic mTLS including certificate rotation? I figure that given we introduce Contour as alternative every component would have to do mTLS on its own again. At least Contour as an option wouldn't be very attractive if the consequence would be unencrypted traffic between CF components.
Given that range of consequences, shouldn't this be an ADR on cf-for-k8s and discussed with the broader community?

Reply:
cc @dieucao

Reply:
Hey @loewenstein, you are correct. This is ADR is currently a draft and we are working on polishing it a bit more, but we will definitely share this with the community.

Reply:
Great. The suggestion to make it an ADR on cf-for-k8s instead of cf-k8s-networking, if it has consequences on cf-for-k8s as a whole, was serious.

Reply:
Trust me, we're all on the same page about the ramifications of this decision. Thanks for tagging Dieu, I'll tag some more folks who might have more background to on this decision, and I hope they'll come up with something official to present to the community at some point:

@ryantang @loganlee @shalako @keshav-pivotal @rosenhouse

Reply:
Yes, this proposal is about creating optionality for those installing CF. Contour is lighter weight than Istio and will have a different feature set. 

You're absolutely right that systems choosing Contour as its ingress controller will forgo some features provided by sidecar proxies. And this is a trade-off those installing CF can make.

Speaking for myself, I believe that modularity is one of the key reasons why Kubernetes has become so popular and powerful. A cynical view (which I disagree with) is that this proposal offers only a less feature rich and therefore inferior alternative. My perspective is that while this proposal seeks to allow two possible `ingress_solution_provider`s today, it creates the possibility for many others in the future. This ADR proposes an investment in the long-term.

Reply:
Thanks @ryantang. I don't take that cynical view either and I agree that modularity is a good thing.

What I am more worried about is the point in time and consequences for other components. And that this is being discussed in cf-k8s-networking instead of cf-for-k8s.
With the upcoming 1.0 release of cf-for-k8s, cf@k8s has not yet reached feature parity with cf@bosh. There are things missing inside and outside of networking (e.g. remote buildpacks, route services, and others). Implementing ADR 18 will take investment from networking and - I suppose - from all other component to get mTLS back in place.
So my question is, whether it is the right point in time.

Reply:
@loewenstein it sounds like you're talking about the broader cf-for-k8s feature roadmap, timeline, and relative prioritization. I am certainly not the expert in this area, so I'll let those with more context than I chime in on this topic.

Reply:
I've reached out to Jan for a higher bandwidth conversation; we're coordinating scheduling.